csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs

```diff
@@ -0,0 +1,145 @@
+
+namespace Google.Protobuf
+{
+    /// <summary>
+    /// Provides unsafe operations for working with <see cref="RepeatedField{T}"/> instances. These
+    /// methods are referred to as "unsafe" due to the fact that they potentially expose the backing
+    /// buffer of a <see cref="RepeatedField{T}"/> to the application.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The methods in this class should only be called if it is guaranteed that the
+    /// <see cref="RepeatedField{T}"/> will not be modified while the returned <see cref="Span{T}"/>
+    /// is in use. Modifying a <see cref="RepeatedField{T}"/> while a <see cref="Span{T}"/> is in
+    /// use can lead to unexpected and undesirable consequences in your application, and will likely
+    /// be difficult to debug. Proceed with caution!
+    /// </para>
+    /// <para>
+    /// Modifying a <see cref="RepeatedField{T}"/> while a <see cref="Span{T}"/> is in use can have
+    /// a number of significant side effects:
+    /// </para>
+    /// <list type="bullet">
+    /// <item>
+    /// <description>
+    /// The <see cref="Span{T}"/> may or may not reflect the changes made to the
+    /// <see cref="RepeatedField{T}"/>.
+    /// </description>
+    /// </item>
+    /// <item>
+    /// <description>
+    /// If the backing array of the <see cref="RepeatedField{T}"/> is reallocated, further updates
+    /// to the <see cref="Span{T}"/> will not be visible.
+    /// </description>
+    /// </item>
+    /// <item>
+    /// <description>
+    /// The state of the <see cref="Span{T}"/> is undefined after any modifications to the <see
+    /// cref="RepeatedField{T}"/>.
+    /// </description>
+    /// </item>
+    /// </list>
+    /// </remarks>
+    [SecuritySafeCritical]
+    public static class UnsafeCollectionOperations
+    {
+        /// <summary>
+        /// Returns a <see cref="Span{T}"/> that wraps the backing array of the specified
+        /// <see cref="RepeatedField{T}"/>. The returned <see cref="Span{T}"/> should not be used
+        /// after any modifications are made to the <see cref="RepeatedField{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The element type of the <see cref="RepeatedField{T}"/>.
+        /// </typeparam>
+        /// <param name="field">
+        /// The <see cref="RepeatedField{T}"/> to wrap.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Span{T}"/> that wraps the backing array of the
+        /// <see cref="RepeatedField{T}"/>.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Span<T> AsSpan<T>(RepeatedField<T> field)
+        {
+            Span<T> span = default;
+            if (field is not null)
+            {
+                var size = field.count;
+                var items = field.array;
+
+                Debug.Assert(items is not null, "Implementation depends on RepeatedField<T> always having an array.");
```

Comment:
Please use a normal if/throw rather than Debug.Assert. (We don't use Debug.Assert in any of the production code in protobuf.)

However, I don't think we actually need to check this anyway. I don't think the array can ever be null, and if it ever *were* null it would cause other issues.

Reply:
resolved